### PR TITLE
Tendermint package with proper steps names

### LINKED ISSUE #79: Create a new consensus package called Tendermint using IBFT implementation and rename component respectively
Rename IBFT references to Tendermint and changes the names of the stages:
- `Preprepare -> Proposal`.
- `Prepare -> Prevote`.
- `Commit -> Precommit`.
- `FinalCommited -> Commit`.

### PR DESCRIPTION
**Pull request by JekaMas**
closes #79

### CODE REVIEW

**Comment by JekaMas** [comment]
@yazzaoui @aneequesafdar just a friendly reminder about the CR

**Comment by aneequesafdar** [comment]
I think it might be on standby until the branching strategy is sorted.

**Comment by JekaMas** [comment]
It's easy to change the target branch anytime. It won't change the PR's code at all.

File: consensus/tendermint/core/roundstate.go [hunk 1]
```diff
@@ -27,13 +27,13 @@ import (
...
-// we need to keep a reference of preprepare in order to propose locked proposal when there is a lock and itself is the proposer
-func newRoundState(view *tendermint.View, validatorSet tendermint.ValidatorSet, lockedHash common.Hash, preprepare *tendermint.Preprepare, pendingRequest *tendermint.Request, hasBadProposal func(hash common.Hash) bool) *roundState {
+// lockedHash and proposal are for round change when lock exists,
+// we need to keep a reference of proposal in order to propose locked proposal when there is a lock and itself is the proposer
+func newRoundState(view *tendermint.View, validatorSet tendermint.ValidatorSet, lockedHash common.Hash, proposal *tendermint.Proposal, pendingRequest *tendermint.Request, hasBadProposal func(hash common.Hash) bool) *roundState {
 	return &roundState{
 		round:          view.Round,
 		sequence:       view.Sequence,
-		Preprepare:     preprepare,
+		proposal:       proposal,
```
Comment (aneequesafdar) [inline consensus/tendermint/core/roundstate.go]: What is the reason to make the proposal private?

File: consensus/tendermint/types.go [hunk 2]
```diff
@@ -40,12 +40,12 @@ type Proposal interface {
 }
 
 type Request struct {
-	Proposal Proposal
+	Proposal ProposalBlock
```
Comment (aneequesafdar) [inline consensus/tendermint/types.go]: Why not keep it as Proposal? with just Proposal Block we lose the View (also needs to be renamed/changed) associated with the request.

File: consensus/tendermint/core/roundstate.go (hunk 1 above)
Comment (JekaMas) [reply consensus/tendermint/core/roundstate.go]: To nit to have a name conflict with Proposal method

File: consensus/tendermint/types.go (hunk 2 above)
Comment (JekaMas) [reply consensus/tendermint/types.go]: Because we now have a Proposal struct - that one with View.

File: consensus/tendermint/core/core.go [hunk 3]
```diff
@@ -241,18 +241,18 @@ func (c *core) startNewRound(round *big.Int) {
...
 	if roundChange && c.isProposer() && c.current != nil {
 		// If it is locked, propose the old proposal
 		// If we have pending request, propose pending request
 		if c.current.IsHashLocked() {
 			r := &tendermint.Request{
-				Proposal: c.current.Proposal(), //c.current.Proposal would be the locked proposal by previous proposer, see updateRoundState
+				Proposal: c.current.Proposal().Proposal, //c.current.Proposal would be the locked proposal by previous proposer, see updateRoundState
 			}
-			c.sendPreprepare(r)
+			c.sendProposal(r)
```
Comment (aneequesafdar) [inline consensus/tendermint/core/core.go]: With regards to this [comment](https://github.com/clearmatics/autonity/pull/90#discussion_r281587048) we can see that sending a proposal takes in a Request which now only has ProposalBlock. Thus when a proposal is received we lose the information about when the block was proposed and we need View to say which height and sequence the block was first proposed.

File: consensus/tendermint/core/roundstate.go [hunk 4]
```diff
@@ -85,23 +85,23 @@ func (s *roundState) Subject() *tendermint.Subject {
...
+func (s *roundState) SetProposal(proposal *tendermint.Proposal) {
 	s.mu.Lock()
 	defer s.mu.Unlock()
 
-	s.Preprepare = preprepare
+	s.proposal = proposal
 }
 
-func (s *roundState) Proposal() tendermint.Proposal {
+func (s *roundState) Proposal() *tendermint.Proposal {
```
Comment (aneequesafdar) [inline consensus/tendermint/core/roundstate.go]: Why did you make the Proposal a pointer?

File: consensus/tendermint/core/types.go [hunk 5]
```diff
@@ -33,16 +33,16 @@ type State uint64
 
 const (
 	StateAcceptRequest State = iota
-	StatePreprepared
+	StateProposald
```
Comment (aneequesafdar) [inline consensus/tendermint/core/types.go]: I think we need a better name for this.

File: consensus/tendermint/types.go (hunk 2 above)
Comment (aneequesafdar) [reply consensus/tendermint/types.go]: You are right, I became very confused by the names, too many proposals with different meanings.

File: consensus/tendermint/core/core.go (hunk 3 above)
Comment (aneequesafdar) [reply consensus/tendermint/core/core.go]: Please ignore, as above.

File: consensus/tendermint/core/types.go (hunk 5 above)
Comment (JekaMas) [reply consensus/tendermint/core/types.go]: You're right. We need a better naming. Will try to find it.

File: miner/worker.go [hunk 6]
```diff
@@ -267,10 +267,10 @@ func (w *worker) pendingBlock() *types.Block {
 // start sets the running status as 1 and triggers new work submitting.
 func (w *worker) start() {
 	atomic.StoreInt32(&w.running, 1)
-	if istanbul, ok := w.engine.(consensus.Istanbul); ok {
+	if istanbul, ok := w.engine.(consensus.PoS); ok {
```
Comment (aneequesafdar) [inline miner/worker.go]: This should also be renamed to `pos`

File: consensus/tendermint/backend/engine_test.go [hunk 7]
```diff
@@ -141,7 +141,7 @@ func makeBlock(chain *core.BlockChain, engine *backend, parent *types.Block) (*t
 
 func makeBlockWithoutSeal(chain *core.BlockChain, engine *backend, parent *types.Block) (*types.Block, error) {
 	header := makeHeader(parent, engine.config)
-	engine.Prevote(chain, header)
+	engine.Prepare(chain, header)
```
Comment (aneequesafdar) [inline consensus/tendermint/backend/engine_test.go]: Seems to have been reverted

File: consensus/tendermint/backend/engine_test.go [hunk 8]
```diff
@@ -160,12 +160,12 @@ func makeBlockWithoutSeal(chain *core.BlockChain, engine *backend, parent *types
 func TestPrevote(t *testing.T) {
 	chain, engine := newBlockChain(1)
 	header := makeHeader(chain.Genesis(), engine.config)
-	err := engine.Prevote(chain, header)
+	err := engine.Prepare(chain, header)
```
Comment (aneequesafdar) [inline consensus/tendermint/backend/engine_test.go]: Seems to have been reverted

File: consensus/tendermint/backend/engine_test.go [hunk 9]
```diff
@@ -160,12 +160,12 @@ func makeBlockWithoutSeal(chain *core.BlockChain, engine *backend, parent *types
...
 	if err != nil {
 		t.Errorf("error mismatch: have %v, want nil", err)
 	}
 	header.ParentHash = common.BytesToHash([]byte("1234567890"))
-	err = engine.Prevote(chain, header)
+	err = engine.Prepare(chain, header)
```
Comment (aneequesafdar) [inline consensus/tendermint/backend/engine_test.go]: Seems to have been reverted

File: consensus/tendermint/backend/engine.go [hunk 10]
```diff
@@ -539,10 +539,10 @@ func (sb *backend) retrieveValidators(header *types.Header, parents []*types.Hea
 
 	if len(parents) > 0 {
 		parent := parents[len(parents)-1]
-		var istanbulExtra *types.IstanbulExtra
-		istanbulExtra, err = types.ExtractIstanbulExtra(parent)
+		var tendermintExtra *types.IstanbulExtra
+		tendermintExtra, err = types.ExtractIstanbulExtra(parent)
```
Comment (aneequesafdar) [inline consensus/tendermint/backend/engine.go]: Might be a good idea to change `ExtractIstanbulExtra` to `ExtractPOSExtra`

File: miner/worker.go (hunk 6 above)
Comment (JekaMas) [reply miner/worker.go]: Do you mean 'POS'?

File: consensus/tendermint/backend/engine_test.go (hunk 8 above)
Comment (JekaMas) [reply consensus/tendermint/backend/engine_test.go]: see above

File: consensus/tendermint/backend/engine.go (hunk 10 above)
Comment (JekaMas) [reply consensus/tendermint/backend/engine.go]: A good idea

**Pull request merged**